src/components/Uploader/index.js

```diff
@@ -0,0 +1,222 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import Icon from '../Icon';
+import IconButton from '../IconButton';
+import Spinner from '../Spinner';
+import UploaderImageItem from '../UploaderImageItem';
+import UploaderItem from '../UploaderItem';
+
+import State from './State';
+import { Container, InputContainer, Preview, PreviewActions } from './styles';
+import { STATES } from './constants';
+
+import { toBase64 } from '../../utils';
+
+const texts = {
+  add_document: {
+    id: 'flamingo.file-uploader.add-document',
+    defaultText: 'Click to add a document',
+  },
+  error: {
+    id: 'flamingo.file-uploader.error-title',
+    defaultText: 'Upload failed... Click to try again',
+  },
+  upload_more_files: {
+    id: 'flamingo.file-uploader.upload-more-files',
+    defaultText: 'Add more files',
+  },
+  uploading: {
+    id: 'flamingo.file-uploader.uploading',
+    defaultText: 'Uploading...',
+  },
+};
+
+const isImage = file => /\.(jpe?g|png|gif)$/i.test(file.name);
+
+const Uploader = React.forwardRef(
+  (
+    {
+      id,
+      onChange,
+      translate,
+      multiple,
+      overrides,
+      state: initialState,
+      value,
+      ...props
+    },
+    ref,
+  ) => {
+    const [state, setState] = React.useState(initialState);
+    const [files, setFiles] = React.useState(value);
+    const [preview, setPreview] = React.useState();
+
+    const onClear = () => {
+      setFiles([]);
+      setPreview();
+      setState(STATES.IDLE);
+    };
+
+    const onDeleteFiles = file => {
+      const remainingFiles = files.filter(({ name }) => name !== file.name);
+
+      if (remainingFiles.length === 0) {
+        onClear();
+        return;
+      }
+
+      setFiles(remainingFiles);
+      onChange(remainingFiles);
+    };
+
+    const onMultipleFilesChange = inputFiles => {
+      // convert event.target.result
+      // into an iterable array of files
+      const filesArr = [...inputFiles];
+      const nextFiles = [...files, ...filesArr];
+
+      setFiles(nextFiles);
+      setState(STATES.FILES);
+    };
+
+    const onSingleFileChange = async inputFile => {
+      if (!isImage(inputFile)) {
+        onMultipleFilesChange([inputFile]);
+        return;
+      }
+
+      setPreview(await toBase64(inputFile));
+      setState(STATES.SINGLE_IMAGE);
+    };
+
+    const onFilesChange = inputFiles => {
+      setState(STATES.LOADING);
+      onChange(inputFiles);
+
+      if (inputFiles.length === 1) onSingleFileChange(inputFiles[0]);
```

Comment:
Instead of going into singleFileChange to check if it's an image just to execute multipleFiles again, why not do it here directly instead? 

```suggestion
      if (inputFiles.length === 1 && isImage(inputFile)) onSingleFileChange(inputFiles[0]);
```

Reply:
Good point 👍

Reply:
Done in https://github.com/heetch/flamingo/pull/171/commits/582d8e84c22391e590dbcbe9f7f2083d71aa01e1
WDYT @obedparla?

Reply:
Nevermind, I merged both fn into a single one, it's even better 😄 
https://github.com/heetch/flamingo/pull/171/commits/4a93e74ca25e18f2fb2ce24576cdf1b101cb07a4

Reply:
Looks great 👍